hca/dss/__init__.py

```diff
@@ -1,172 +1,139 @@
+from .. import logger
+from .upload_to_cloud import upload_to_cloud
+
+class DSSClient(SwaggerClient):
+    """
+    Client for the Data Storage Service API.
+    """
+    UPLOAD_BACKOFF_FACTOR = 1.618
+    def __init__(self, *args, **kwargs):
+        super(DSSClient, self).__init__(*args, **kwargs)
+        self.commands += [self.download, self.upload]
+
+    def download(self, bundle_uuid, replica, version="", dest_name=""):
+        """
+        Download a bundle and save it to the local filesystem as a directory.
+        """
+        if not dest_name:
+            dest_name = bundle_uuid
+
+        bundle = self.get_bundle(uuid=bundle_uuid, replica=replica, version=version if version else None)["bundle"]
+
+        if not os.path.isdir(dest_name):
+            os.makedirs(dest_name)
+
+        for file_ in bundle["files"]:
+            file_uuid = file_["uuid"]
+            filename = file_.get("name", file_uuid)
+
+            logger.info("File %s: Retrieving...", filename)
+            session = self.get_session()
+            session.stream = True
+            response = self.get_file._request(dict(uuid=file_uuid, replica=replica))
+
+            try:
+                if response.ok:
+                    file_path = os.path.join(dest_name, filename)
+                    logger.info("%s", "File {}: GET SUCCEEDED. Writing to disk.".format(filename))
+                    with open(file_path, "wb") as fh:
+                        for chunk in response.iter_content(chunk_size=1024*1024):
+                            if chunk:
+                                fh.write(chunk)
+                    logger.info("%s", "File {}: GET SUCCEEDED. Stored at {}.".format(filename, file_path))
+
+                else:
+                    logger.error("%s", "File {}: GET FAILED.".format(filename))
+                    logger.error("%s", "Response: {}".format(response.text))
+            finally:
+                response.close()
+                session.stream = False
+        return {}
+
+    def upload(self, src_dir, replica, staging_bucket, timeout_seconds=1200):
+        """
+        Upload a directory of files from the local filesystem and create a bundle containing the uploaded files.
+
+        This method requires the use of a client-controlled object storage bucket to stage the data for upload.
+        """
+        bundle_uuid = str(uuid.uuid4())
+        files_to_upload, files_uploaded = [], []
+        for filename in os.listdir(src_dir):
+            full_file_name = os.path.join(src_dir, filename)
+            files_to_upload.append(open(full_file_name, "rb"))
+
+        logger.info("Uploading %i files from %s to %s", len(files_to_upload), src_dir, staging_bucket)
+        file_uuids, uploaded_keys = upload_to_cloud(files_to_upload, staging_bucket=staging_bucket, replica=replica,
+                                                    from_cloud=False)
+        for file_handle in filter(lambda x: not isinstance(x, str), files_to_upload):
```

Comment:
It used to be that this could be an S3 URL prefix, but with this refactor it's no longer possible to pass that in. Thanks for the catch, I've removed the filter.